agent/util-scripts/pbench-tool-meister-stop

```diff
@@ -42,6 +38,88 @@ def is_running(pid):
     return True
 
 
+def wait_for_pid(pid):
+    """wait_for_pid - wait for a process to actually stop running.
+    """
+    while is_running(pid):
+        time.sleep(0.1)
+
+
+def graceful_shutdown(
+    benchmark_run_dir, full_hostname, group, redis_server_pid_file, logger
+):
+    # The assumption/assertion here is that the tool meister "stop" command is
+    # run on the same node as the tool meister "start" command ran, creating
+    # the local Tool Data Sink and the optional local Tool Meister. We want to
+    # make sure anything "local" to this stop command is shut down gracefully
+    # before we report back to the user.  If Tool Meisters from remote nodes
+    # have already reported that they have received the "terminate" message,
+    # then we trust they will shutdown gracefully themselves.
+    try:
+        tds_pid_file = benchmark_run_dir / "tm" / "pbench-tool-data-sink.pid"
+        try:
+            pid_str = tds_pid_file.read_text()
+        except OSError as exc:
+            if exc.errno != errno.ENOENT:
+                raise
+        else:
+            tds_pid = int(pid_str)
+            logger.debug("waiting for tool-data-sink (%d) to exit", tds_pid)
+            wait_for_pid(tds_pid)
+    except Exception:
+        logger.exception("Exception encountered waiting for tool-data-sink")
+        ret_val = 1
+    else:
+        ret_val = 0
+
+    try:
+        ltm_pid_file = benchmark_run_dir / "tm" / f"tm-{group}-{full_hostname}.pid"
+        try:
+            pid_str = ltm_pid_file.read_text()
+        except OSError as exc:
+            if exc.errno != errno.ENOENT:
+                raise
+        else:
+            ltm_pid = int(pid_str)
+            logger.debug("waiting for local tool-meister (%d) to exit", ltm_pid)
+            wait_for_pid(ltm_pid)
+    except Exception:
+        logger.exception("Exception encountered waiting for local tool-meister")
+        ret_val = 1
```

Comment:
This looks ripe for refactoring, given how similar it is to the previous and subsequent blocks.

Reply:
Tracked in https://github.com/distributed-system-analysis/pbench/issues/2175.